Draftail updates, refactors, bug fixes 2

thibaudcolas opened the pull request:
Follow-up to #4217. Addresses the following Draftail tasks:

- [x] Extending Draftail docs
- [x] Update editor guide
- [x] Extending Hallo docs update
- [x] Impossible to use SVG icons from within Wagtail
- [x] #4227
- [x] Fix Markdown shortcuts for blocks removing styles and entities at the end of the block text
- [x] Toolbar buttons spacing & alignment in Firefox, Safari

Documents some more changes from the Draftail PR:

* Remove Hallo editor ``.richtext`` CSS class in favour of more explicit extension points (Thibaud Colas)
* Expose React-related dependencies as global variables for extension in the admin interface (Thibaud Colas)
* Fix: Only load Hallo CSS if Hallo is in use (Thibaud Colas)
* Fix: Prevent style leak of Wagtail panel icons in widgets using ``h2`` elements (Thibaud Colas)

Also some refactorings:

- Hallo docs are now in a separate page.
- Changes the Draftail client-side API to register plugins to be more similar to `register_rich_text_feature`.
- Updates Draftail and Draft.js to their latest versions (see the separate CHANGELOGS).
- 100% test coverage for client-side Draftail code.


----

# Remaining tasks

<details>

- [ ] The background colour flickers when the link edit takes focus
- [ ] Editor crash when inserting new link, document, image embed in IE11 (specific to Wagtail)
- [ ] Toolbar does not move in Safari

## After merge

### Bugs

### Refactorings

### Documentation

### Postponed

- [ ] Support tooltip rendering top-right when necessary (depending on available space).
- [ ] Distinguish between internal and external links in the editor by using different icons.
- [ ] Finish "Edit" flow for images and embeds #2674.
- [ ] Re-add `alt` editing for image block.
- [ ] "Link text" field in page chooser should be pre-filled with text
- [ ] Opening up a dialog (eg. new link, document, image) causes the editable text area to go white, then the dialog loads. However, on a slow connection (mine right now), this may take up to two seconds or more. The text going white and becoming non-editable is a bit confusing - it might be good to add a loading indicator? Or maybe the screen should go darker straight away (ie. before the modal response returns). My connection is quite slow so not sure what the best way around this is. **Should be addressed in the choosers, rather than the editor**.
- [ ] Minor nitpick: If you have a lot of content in the editor (say two pages worth), and you are editing something towards the end of the content, then you clear that content. Your scroll position is kept to further down the editor page, and the draftail editor essentially disappears. This may not be worth solving - but wanted to mention it.
- [ ] Editor behavior when image / embed / link / document don't have data defined (remove the entity/block?)
- [ ] Order of the entities in the toolbar

### Tests

- [ ] Hunt for more content conversion issues

### Nice to haves

- [ ] Image entity alt text save on "Enter"
- [ ] External/email link chooser isn't initialised with currently selected text
- [ ] (After React 16 upgrade has been merged) Add error boundary around editor component

### Cleanup

- [ ] Change all Draftail code to 2-space indent to match Wagtail
- [ ] Move Hallo JS code to a Webpack entry as well, and clean up code.
- [ ] Move modal workflow code to Webpack (and cleanup).
- [ ] Fix / remove / adapt class names defined for image formats
- [ ] Default max list nesting, and configurability
- [ ] Decorators configurability
- [ ] Deprecate images and embeds in rich text? (in favour of Streamfield)

### Other bugs

- [ ] `cursor: pointer;` on form-field `.add` `p`.

</details>

# Fixed

<details>

### Documentation

- [x] Usage
- [x] Extending Draftail docs
- [x] Update editor guide
- [x] Extending Hallo docs update
- [x] Draftail known issues https://github.com/springload/draftail/issues/138
- [x] Release notes

### UI

- [x] Improve support for custom icons
- [x] Make all entity-related components definitions in Draftail integrations rather than in core
- [x] Move tooltip outside of Draftail core.
- [x] Implement latest UI iteration based on Bristol sprint designs
- [x] Move latest version of Wagtail-only components into Wagtail
- [x] Adapt UI styles to coexist with Hallo
- [x] Review latest UI iteration within Wagtail
- [x] Post-review fixes and potential release
- ~[ ] Move nested lists styles over~
- [x] Reproduce Hallo's "focus when adding a sequence member" in sequence.js
- [x] Reproduce Hallo field styles (focus)
- [x] i18n
- [x] Toolbar icons alignment
- [x] Toolbar border radius consistency
- [x] Tooltip background color flicker
- [x] Draftail editor padding/spacing
- [x] Toolbar tooltip obstructing first lines of content
- [x] Color contrast of the 'remove' button is too low
- [x] Left and right paddings of 'edit' and 'remove' buttons should match the 'save draft' button. The button paddings do not collapse right on mobile viewport
- [x] Support absence of data on Link
- [x] Support absence of data on Document
- [x] Support absence of data on Image
- [x] Support absence of data on Embed

### Build

- [x] Expose Wagtail dependencies as globals
- [x] Ensure Draft.js vendor deps are only loaded when Draftail is.

### Entities 

- [x] Updating image alt text / alignment through the inline form is laggy - changes only get committed to the underlying contentstate data when the rich text field is refocused, and only propagate back to the form after an edit is made to the rich text
- [x] Make sure sources handle creation & edit (link, document)
- [x] Make sure sources handle creation & edit (image, embed). Edit: Considering removing the "Edit" button for now.
- [x] Investigate scroll on focus
- [x] Handle link text in link source (`prefer_this_title_as_link_text: true`, `title: "text"`)
- [x] Make sure image formats are configurable via image chooser
- [x] Handle server errors without locking the editor

### Other

- [x] Finish first draft of styling/theming API
- [x] Link icon check based on `mailto` `startsWith`
- [x] Finish rich text copy-paste filtering https://github.com/springload/draftail/issues/123
- [x] Replace constants / hard-coded strings in wagtail_hooks, rich_text, contentstate.py
- [x] Filter pasted `tel` links. Ideally, turn pasted link filtering into a [real whitelist](https://github.com/wagtail/wagtail/pull/4136#discussion_r163582141).

### Bugs

- ~[ ] No-break text overflow~
- ~[ ] Prevent toolbar from obstructing last block~
- [x] Impossible to use SVG icons from within Wagtail
- [x] Toolbar buttons spacing & alignment in Firefox, Safari

### Testing

- [x] Cross-browser test within Wagtail
- [x] More tests on mobile devices, in Wagtail
- [x] Check pasting from Hallo to Draftail
- [x] Check pasting from more recent versions of Word https://github.com/springload/draftail/issues/123
- [x] 100% unit test coverage
- ~[ ] Troubleshoot ContentState conversion bug for "HTML-sanitised" text.~

### Nice to haves

- ~[ ] Add CSS linting to Draftail https://github.com/springload/draftail/issues/126~
- [x] Fix block type reset annoyance https://github.com/springload/draftail/issues/105
- ~[ ] Automated integration tests https://github.com/springload/draftail/issues/12~
- [x] i18n for horizontal line, line break, undo-redo
- [x] Re-check known issue https://github.com/springload/draftail/issues/108
- [x] `LinkElementHandler` -> `InlineEntityElementHandler` with configurable `mutability`.
- [x] Automatically-generated depth styles for list items

### Refactorings

- [x] Remove default {} when finishing https://github.com/springload/wagtaildraftail/issues/32
- [x] Fix silly `ugettext_lazy` serialisation
- [x] Add `data-` attribute to Hallo editor to define documented integration point for plugins.
- [x] Make editor `font-family` and `font-size` overridable via Sass variable
- [x] Remove unused Sass variables
- [x] Namespace variables and mixins
- [x] Stop loading Hallo CSS even if Hallo isn't loaded
- [x] `user-select: none;` on button labels
- ~[ ] Move block spacing styles to Draftail~
- [x] Add unknown attribute (`id`) filter to Draft.js filters
- [x] Refactor sources composition over inheritance
- [x] Remove unused `.richtext-image img`, `.richtext-image figcaption`

### Decisions

- [x] (soft) line break – enabled or disabled by default?
- [x] undo redo controls displayed in toolbar?

### Cross-browser bugs

- [x] Media block bottom border in Safari, Android Chrome
- [x] Line break button renders badly in Edge
- [x] Placeholder misaligned in ~iOS Safari~ full-width fields

### Other bugs

- ~[ ] Using a custom serif font makes the text flicker when loading bold / italic variants~
- [x] `LOADING: "{% trans 'Loading...' %}",` translation not using ellipsis char. #4218 

</details>

gasman commented:
Thanks @thibaudcolas!

I think it's about time we split "extending hallo.js" and "extending Draftail" into their own sub-pages of advanced_topics/customisation - they're outgrowing the list of topics on the 'page editing interface' page (and are on another level of complexity to the other topics covered there), and splitting them up means we don't have to struggle with the "you can skip over this section..." signposting. I suggest we stop after the list of built-in features, with a note like: "this list can be extended, and the process for doing so depends on the editor you're using: see [Draftail] or [hallo.js]".

thibaudcolas commented:
@gasman this is good to go. I split the docs in two pages, and added an in-depth example about custom entities. Custom block-level entities (image, embed) are left out in the docs, specifically referring people to StreamField instead.

Since your last review, I also did some refactoring, added more tests, and fixed issues #4227 and the last problem from #4222 (datepicker z-index vs toolbar z-index). This also fixes #4221.

Finally, there is a Draftail + Draft.js upgrade in there. I wasn't too sure the Draft.js upgrade was worth it since this is already in beta, but there are a few great bugfixes in there so I added it after ample (hopefully enough) testing.

---

I wouldn't mind doing a third round of changes before an RC. Specifically, to address:

> 1. The background colour flickers when the link edit takes focus.
> 2. Toolbar does not move in Safari
> 3. Editor crash when inserting new link, document, image embed in IE11 (specific to Wagtail)

I know the cause of all 3. 1 and 2 aren't high priority but should be straightforward fixes. 3 is harder to troubleshoot but IMHO important to get in. I know it is due to some adverse interaction between the ModalWorkflow's modal and `contenteditable`, presumably related to focus management.
